Add HMAC multi-mode support (SHA-2 384/512 digest sizes and variable key lengths)

AlexJones0 opened the pull request:
This PR completes the remaining work to get Earlgrey's HMAC to match the latest hardware functionality. May be easier to review commit-by-commit. See commit messages for more details.

Adds support to the HMAC for variable digest sizes (SHA-2 256/384/512) and key lengths (128/256/384/512/1024 bits) by setting the corresponding fields in the config register. Related checks are added, and existing functionality replaced which assumed SHA-2 256 & key size <= 512 bits. Also cleans up the last minor TODO.

Tested against existing passing tests: `hmac_enc_test`, `hmac_sha256_functest`, `sha256_functest`, `hmac_smoketest`, `wots_test`, `verify_test_hardcoded`, `fors_test`, etc.
Also tested to now make previously failing tests pass: `hmac_functest`, `sha384_functest`, `hmac_sha384_functest`, `sha512_functest` and `hmac_sha512_functest`.
All tests were built from [master](https://github.com/lowrisc/opentitan/commits/37fc765a852e618a202a40d24a33659cfe449580) on OpenTitan.

AlexJones0 committed:
[ot] hw/opentitan: ot_hmac: Implement key/digest size regs & checks

This commit implements initial register interface support for the key
length and digest size fields in the config register of OpenTitan's
HMAC. It checks that the values written are valid one-hot encodings, and
maps them as is done in current hardware if not.

This also implements some additional checks that the combination of (key
length, digest size, HMAC enable) is valid when attempting to run the
START/CONTINUE command, otherwise producing a (new type of) software
error.

Also fixes some comment formatting (code style).

AlexJones0 committed:
[ot] hw/opentitan: ot_hmac: Make HMAC outer hash independent of key/digest size

This commit updates the size of the key and digest registers to their
new maximum sizes in support of HMAC Multi-Mode (SHA-2 256/384/512
Engine), which is 64 bytes for the digest and 128 bytes for the key.

The current HMAC logic which processes the input padding and performs
the outer hash (using the outer padding) by default assumes that the
key length is 64 bytes (512 bits), and that you are hence using the
entire set of registers. They also assume that the digest is 32 bytes
(256 bits), and that you are again hence using the entire register set.

The new logic now uses the key_length and digest_size config register
fields to determine the exact padding and digest sizes that should be
used in the HMAC calculations.

AlexJones0 committed:
[ot] hw/opentitan: ot_hmac: Add HMAC SHA-2 384/512 & key size support

This commit introduces full support for using the SHA-2 384 and SHA-2
512 algorithms for hashing alongside the existing SHA-2 256 algorithm,
by setting the `digest_size` field in the configuration register. It
likewise adds support for using a range of key sizes (128/256/384/512/
1024), and not just always using a 512 bit key - this is set in the
`key_length` field in the configuration register.

The existing tomcrypt library is used to support the additional hash
functionality, which uses `sha512` functionality for both sha384 and
sha512 in its implementation.

This commit also fixes an issue previously introduced where two of
the `bswap`s on the digest were the wrong way around, cancelling
each other out but technically producing an incorrect intermediary
result. This fixes that behaviour, and updates the documentation to
make the reasoning behind each endian swap clear.

AlexJones0 committed:
[ot] hw/opentitan: ot_hmac: Remove remaining unnecessary TODO

This removes the last remaining TODO from `ot_hmac` (after the
multi-mode SHA384/SHA512 support is added). Looking at the RTL,
it wipes the secret key regardless of whether the engine is
active or not, even if mid-operation. So this TODO can be
safely removed to mimic the real hardware.

Also fixes a comment line length issue from a previous commit.

AlexJones0 committed:
[ot] hw/opentitan: ot_hmac: Digest size write should not affect current hash

This commit updates the functionality for the HMAC's digest size
configuration register field, updating it to match the current hardware
behaviour. The hardware uses a `digest_size_started` field to allow the
user to write a digest size at any time, while ensuring that if a hash
is in progress it will carry on using the existing digest size.

A similar field is introduced into the context/state struct to allow the
QEMU HMAC model to keep track of the digest size when the last
START/CONTINUE command was sent, and use that in place of checking the
register value every time.

rivos-eblot commented on hw/opentitan/ot_hmac.c:
```diff
@@ -222,6 +223,22 @@ static const char *REG_NAMES[REGS_COUNT] = {
+    HMAC_SHA2_256,
+    HMAC_SHA2_384,
+    HMAC_SHA2_512,
+    HMAC_SHA2_NONE,
```
It would be better to use None as the first value (0), as long as it does not have a strong tie with a HW value.
Other QEMU OT IPs try to use this rule.

AlexJones0 replied:
In this case I was just following the order of the definitions in the documentation, which matches the bit order of the one-hot encoding (i.e. 0x1u, 0x2u, 0x4u, etc.) AFAIK there is no other specific strong ties to a HW value so I'll change this now to conform to other QEMU OT IPs.

rivos-eblot commented on hw/opentitan/ot_hmac.c:
```diff
@@ -222,6 +223,22 @@ static const char *REG_NAMES[REGS_COUNT] = {
 };
 #undef REG_NAME_ENTRY
 
+enum OtHMACDigestSize {
```
If the enum (or structure or union) is used as type (I mean not only its enumerated value, but the enum as a member of a structure or function argument/return type), please define it as a type. It seems to be the convention in QEMU, so we followed it in other IPs (but I pretty sure we can find the opposite rule somewhere in QEMU history :-)), _e.g._
```c
typedef enum OtHMACDigestSize {
 ...
} OtHMACDigestSize;
```
and only use the typedef version later on.

rivos-eblot commented on hw/opentitan/ot_hmac.c:
```diff
@@ -259,6 +277,102 @@ struct OtHMACState {
+    }
+}
+
+static inline size_t ot_hmac_get_digest_bytes(enum OtHMACDigestSize digest_size)
```
maybe let the compiler takes the decision for inlining, since you're using `__func__`? I'm not sure what to do in this case.

rivos-eblot commented on hw/opentitan/ot_hmac.c:
```diff
@@ -326,16 +557,24 @@ static void ot_hmac_compute_digest(OtHMACState *s)
-        for (unsigned i = 0; i < ARRAY_SIZE(opad); i++) {
-            opad[i] ^= 0x5c5c5c5c5c5c5c5cull;
+        memcpy(opad, s->regs->key, key_length_b);
+        for (unsigned i = 0; i < pad_length_w; i++) {
```
i -> `ix` or `idx`

rivos-eblot replied:
beware that `unsigned` is likely 32 bits while `pad_length_w` is likely 64 bits (on 64 bit hosts)

AlexJones0 replied:
I was mostly aiming to leave the existing code as is where possible, but I'll change the name of this variable to match current styling. Based on the context of the actual values I think the type widths should be fine here but as good practice I'll change the `unsigned` to a `size_t`.

rivos-eblot replied:
Please do. HMAC is one of the oldest IP we've implemented, and we learnt and refined the rules since then, and we have not updated the code. As you're actively working on it, this is a nice opportunity to align the code. Sorry for adding some workload here.

rivos-eblot commented on hw/opentitan/ot_hmac.c:
```diff
@@ -259,6 +277,102 @@ struct OtHMACState {
+        return 64u;
+    case HMAC_SHA2_NONE:
+    default:
+        qemu_log_mask(LOG_GUEST_ERROR,
```
I have not tracked down the entire call stack, but it seems this case arrives late after other computations have been performed. Does it behave as the HW actually does?
I would have though that the computation is not started in this case, in which case this case should not happen and `g_assert_not_reached` could be reached? Maybe this check may be performed earlier?

It is always better to use `LOG_GUEST_ERROR` as close to the actual error occurence, likely a register write access.

rivos-eblot replied:
> Unsupported/invalid values and all-zero values are mapped to Key_None. With this value, when HMAC is triggered to start operation (via hash_start or hash_continue), it will be blocked from starting and an error is signalled to the SW. If only unkeyed SHA-2 is configured (hmac_en == 0), starting is not blocked, since this does not require a key.

I just skimmed over the doc but it seems we first need to validate the keylength, and either start the computation or report an error.

AlexJones0 replied:
My changes should already do this, I believe - see [lines 885-900](https://github.com/lowRISC/qemu/blob/348f406e1b611557bbd7275ee1449165787a1906/hw/opentitan/ot_hmac.c#L885-L900). The HMAC in this case should mimic the HW in catching the error as soon as the start or continue command is received, and will report it as a software-visible error. In reality this `qemu_log_mask` is more as a "the QEMU code is obviously wrong if this function gets called with this state, so tell the user". I was not aware of the `g_assert_not_reached`, so I'll replace it with that instead.

AlexJones0 replied:
Along with the other guest errors, this has been changed to a `g_assert_not_reached` call with a comment explaining why that control flow path should be unreachable. The original hardware functionality has been retained as I described above - this is just additional protection within the QEMU model.

rivos-eblot commented on hw/opentitan/ot_hmac.c:
```diff
@@ -259,6 +277,102 @@ struct OtHMACState {
+    case HMAC_SHA2_NONE:
+    default:
+        qemu_log_mask(LOG_GUEST_ERROR,
+                      "%s: Cannot get block size if digest size is not set. \n",
```
Beware of the extra space char (multiple occurences)

rivos-eblot commented on hw/opentitan/ot_hmac.c:
```diff
@@ -259,6 +277,102 @@ struct OtHMACState {
+
+static inline size_t ot_hmac_get_key_bytes(OtHMACState *s)
+{
+    switch (ot_hmac_get_key_length(s->regs->cfg)) {
```
It seems to be a dual dispatch (switch, switch).  As it looks the HW values are always 2^N, so maybe we could use some builtin (I think there are macros in QEMU that map them) to check that only 1 bit is set (`__builtin_popcount`), and shift the valid values to generate the actual keylength / byte count?

(maybe I'm wrong and the HW values do not allow this kind of implementation, please let me know)

AlexJones0 replied:
~The HW should allow for this kind of implementation, as the encoding is one-hot. I'll have a look at if I can do this to avoid the dual dispatch.~

AlexJones0 replied:
Never mind, I completely missed that this is unfortunately not the case - `HMAC_KEY_384` is `48u` bytes, which is not 2^N. If not for that case then I was going to try something like the below, but this is not possible to optimise away (at least using this method):

```c
static size_t ot_hmac_get_key_bytes(OtHMACState *s)
{
    uint32_t key_length = (cfg_reg & R_CFG_KEY_LENGTH_MASK);
    if (ctpop32(key_length) != 1) {
        g_assert_not_reached();
        return 0u;
    }
    static_assert(R_CFG_KEY_LENGTH_SHIFT > 4);
    return key_length >> (R_CFG_KEY_LENGTH_SHIFT - 4);
}
```

rivos-eblot replied:
Ok, I missed it as well.

AlexJones0 replied:
The other alternative is that, for the three functions with dual dispatch (`ot_hmac_get_digest_bytes`, `ot_hmac_get_block_size_bytes` and `ot_hmac_get_key_bytes`), they all have the same cases as their respective `ot_hmac_get_digest_size` and `ot_hmac_get_key_length` functions. So, I could just eliminate the calls and use the values directly, e.g.:

```c
static size_t ot_hmac_get_key_bytes(OtHMACState *s)
{
    switch ((s->regs->cfg & R_CFG_KEY_LENGTH_MASK) >> R_CFG_KEY_LENGTH_SHIFT) {
    case 0x01u:
        return 16u;
    case 0x02u:
        return 32u;
    case 0x04u:
        return 48u;
    case 0x08u:
        return 64u;
    case 0x10u:
        return 128u;
    case 0x02u:
    default:
        g_assert_not_reached();
        return 0u;
    }
}
```
The benefit to this is the micro-optimisation, which may be worth it if you think it is performance critical. The downside is you lose the single sources of truth for the digest size and key length - now if the hardware or implementation changes, you need to update these values in several places. We could replace this by assigning the enums literal values which we then use in the cases. How do you feel about this approach?

rivos-eblot replied:
Note about `static_assert`: I think clang-tidy will complain that a message is required as message-less assertion is C++ only, _e.g._ 
```c
static_assert(cond, "reason for failure");
```

rivos-eblot replied:
> The benefit to this is the micro-optimisation, which may be worth it if you think it is performance critical.

I think it is not worth it, the compiler will likely optimize the switch/case anyway. Prefer the most readable/safe version for now. Thanks

rivos-eblot commented on hw/opentitan/ot_hmac.c:
```diff
@@ -640,17 +914,27 @@ static void ot_hmac_regs_write(void *opaque, hwaddr addr, uint64_t value,
+                the right with 0s. */
+                size_t pad_length_b = MAX(key_length_b, block_size_b);
+                size_t pad_length_w = pad_length_b / sizeof(uint32_t);
+                uint32_t ipad[OT_HMAC_MAX_KEY_LENGTH / sizeof(uint32_t)];
```
Can we have length/size which are not a multiple of 8 bytes? If not, we could use `uint64_t` values here. Just checking :)

AlexJones0 replied:
From my understanding, all lengths/sizes are multiples of 16 bytes, so this can definitely be changed - I mostly used `uint32_t` to match the type width of `s->regs->key`, as it made the logic easier to follow in my head. Is there a reason that you would prefer `uint64_t` here?

rivos-eblot replied:
This is the native size for most hosts nowadays (x86_64, aarch64). Micro optimization but the less time/loops here the faster the results and/or the more time available for all the other IPs in the I/O thread. HMAC implementation is a known bottleneck at the moment on some workloads (such as verifying large FW/data chunks)

AlexJones0 replied:
That makes sense, thanks for the detailed explanation. I'll change that back to using `uint64_t`.

rivos-eblot replied:
Thanks.

rivos-eblot commented on hw/opentitan/ot_hmac.c:
```diff
@@ -612,19 +854,51 @@ static void ot_hmac_regs_write(void *opaque, hwaddr addr, uint64_t value,
+        digest_size = ot_hmac_get_digest_size(val32);
+        if (digest_size == HMAC_SHA2_NONE) {
+            val32 &= ~R_CFG_DIGEST_SIZE_MASK;
+            val32 |= 0x8u << R_CFG_DIGEST_SIZE_SHIFT;
```
Would be nice to use some `#define` constants here (for `SHA2_NONE` and `KEY_NONE`)

AlexJones0 replied:
Can you clarify what exactly you mean here? Are you referring to making manifest constants for the `0x8u << R_CFG_DIGEST_SIZE_SHIFT` and `0x20u << R_CFG_KEY_LENGTH_SHIFT` values, or for something else?

AlexJones0 replied:
Ah, apologies, I think I misunderstood - are you referring to using e.g. `FIELD_DP32(storage, reg, field, val)`?

rivos-eblot replied:
Something like
```c

// somewhere after the REG32 section
#define CFG_DIGEST_SHA2_NONE 0x8u
#define CFG_KEY_LENGTH_NONE  0x20u

// ...

val32 |= CFG_DIGEST_SHA2_NONE << R_CFG_DIGEST_SIZE_SHIFT;
// ...
val32 |= CFG_KEY_LENGTH_NONE << R_CFG_KEY_LENGTH_SHIFT;
```

rivos-eblot replied:
> Ah, apologies, I think I misunderstood - are you referring to using e.g. `FIELD_DP32(storage, reg, field, val)`?

No, `val32 |= ...` is fine.

rivos-eblot commented on hw/opentitan/ot_hmac.c:
```diff
@@ -699,13 +983,11 @@ static void ot_hmac_regs_write(void *opaque, hwaddr addr, uint64_t value,
-            for (unsigned i = 0; i < digest_length; i++) {
-                s->ctx->state.sha256.state[i] = s->regs->digest[i];
-            }
+            /* Hold the previous digest size until the HMAC is started with the
```
Unrelated, can you take the opportunity of this change to fix the comment line width @ 968? Thanks.

rivos-eblot reviewed (changes requested):
Additional note: it might me useful to add lowRISC and your address to the file caption?

AlexJones0 committed:
[ot] hw/opentitan: ot_hmac: Add copyright/author information

AlexJones0 commented:
@rivos-eblot Thank you for the thorough review. I believe I've addressed all the comments now. Please let me know if there are issues with how I've added lowRISC/myself to the file caption - I'm not sure if there are any specific standards or best practices you would prefer for me to follow there. I've re-run the tests described in my initial comment again and they are all still passing with the latest changes.

loiclefort commented on hw/opentitan/ot_hmac.c:
```diff
@@ -151,6 +152,11 @@ REG32(MSG_LENGTH_UPPER, 0xe8u)
 
+/* value representing 'SHA2_NONE' in the config digest size field */
+#define OT_HMAC_CFG_DIGEST_SHA2_NONE 0x8u
+/* value repersenting 'KEY_NONE' in the config key length field */
```
typo: `repersenting` -> `representing`

loiclefort commented on hw/opentitan/ot_hmac.c:
```diff
@@ -407,16 +407,97 @@ static void ot_hmac_writeback_digest_state(OtHMACState *s)
+        s->ctx->state.sha256.curlen = 0;
+        s->ctx->state.sha256.length = s->regs->msg_length;
+        for (unsigned idx = 0; idx < 8u; idx++) {
+            s->ctx->state.sha256.state[idx] = s->regs->digest[idx];
```
Maybe use `LOAD32H` for symmetry with `STORE32H` in `ot_hmac_writeback_digest_state`?

loiclefort commented on hw/opentitan/ot_hmac.c:
```diff
@@ -407,16 +407,97 @@ static void ot_hmac_writeback_digest_state(OtHMACState *s)
+        for (unsigned idx = 0; idx < 8u; idx++) {
+            s->ctx->state.sha512.state[idx] =
+                ((uint64_t)s->regs->digest[idx + 1] << 32u) |
+                s->regs->digest[idx];
```
Same here, use `LOAD64H`?

loiclefort commented on hw/opentitan/ot_hmac.c:
```diff
@@ -407,16 +407,97 @@ static void ot_hmac_writeback_digest_state(OtHMACState *s)
+    case HMAC_SHA2_384:
+        /* Even though SHA384 only uses the first six uint64_t values of
+            the SHA512 digest, we must store them all for intermediary
+            computation. */
```
QEMU coding-style for multi-line comments is:
```
/*
 * line 1
 * line 2
 */
```
(there are several occurrences to fix)

AlexJones0 commented:
Latest force push should resolve all further comments and fix a couple of other minor issues I noticed. I've tested again on the same tests and still see all tests passing.

rivos-eblot commented:
> @rivos-eblot Thank you for the thorough review. I believe I've addressed all the comments now. Please let me know if there are issues with how I've added lowRISC/myself to the file caption - I'm not sure if there are any specific standards or best practices you would prefer for me to follow there.

Hi @AlexJones0,
I'm not sure there are rules for those. I think it would be better to split the copyright line into:
```
* Copyright (c) 2022-2024 Rivos, Inc.
* Copyright (c) 2024 lowRISC contributors.
```
Otherwise, LGTM.

rivos-eblot reviewed (approved):
LGTM
